tools/src/tensil/tools/compiler/Estimator.scala

```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright © 2019-2022 Tensil AI Company */
+
+package tensil.tools.compiler
+
+import tensil.Architecture
+
+class Estimate(initCycles: Long, initEnergy: Long) {
+  private var curCycles = initCycles
+  private var curEnergy = initEnergy
+
+  def cycles = curCycles
+  def energy = curEnergy
+
+  def splitCycles(cyclesToSplit: Long): Estimate = {
+    require(cyclesToSplit < cycles)
+
+    val energyToSplit =
+      ((cyclesToSplit.toFloat / cycles.toFloat) * energy.toFloat).toLong
+
+    curCycles -= cyclesToSplit
+    curEnergy -= energyToSplit
+
+    new Estimate(cyclesToSplit, energyToSplit)
+  }
+}
+
+class Estimator(arch: Architecture) {
+  private var previousOpcode = Opcode.Wait
+  private var previousFlags  = 0
+
+  private val InternalTransferEnergy = 10
+  private val InternalTransferCycles = 1
+
+  private val DRAMTransferEnergy = 100
+
+  private val DRAMTransferSetupReadCycles  = 32
+  private val DRAMTransferSetupWriteCycles = 0
+  private val DRAMTransferReadCycles       = 2
+  private val DRAMTransferWriteCycles      = 1
+  private val DRAMTransferWidthBits        = 128
+
+  def estimateCyclesAndEnergy(
+      currentOp: Int,
+      size: Option[MemoryAddressRaw] = None,
+      flags: Int = 0
+  ): Estimate = {
+    val r = currentOp match {
+      case Opcode.Wait =>
+        val cycles = 1L
+        val energy = 0L
+
+        new Estimate(cycles, energy)
+      case Opcode.MatMul => {
+        val cycles =
+          (if (previousOpcode == Opcode.MatMul)
+             (size.get + 1)
+           else if (previousOpcode == Opcode.LoadWeights)
+             (size.get + 1 + arch.arraySize)
+           else
+             (size.get + 1 + 2 * arch.arraySize))
+
+        val energy =
+          (size.get + 1) * arch.arraySize * arch.arraySize
+
+        new Estimate(cycles, energy)
+      }
+
+      case Opcode.SIMD => {
+        val cycles = 1L
+        val energy = arch.arraySize.toLong
+
+        new Estimate(cycles, energy)
+      }
+
+      case Opcode.LoadWeights => {
+        val cycles = (size.get + 1) * InternalTransferCycles
+        val energy = (size.get + 1) * InternalTransferEnergy
+
+        new Estimate(cycles, energy)
+      }
+
+      case Opcode.DataMove =>
+        if (
+          flags == DataMoveFlags.LocalToDRAM0 ||
+          flags == DataMoveFlags.LocalToDRAM1 ||
+          flags == DataMoveFlags.DRAM0ToLocal ||
+          flags == DataMoveFlags.DRAM1ToLocal
+        ) {
+          /*val transfersPerVector = divCeil(
```

Comment:
is this intended to be commented out?

Reply:
Yes. I am planning to return to estimation with idea of re-using runtime traces. I want to keep this elaborate version of estimation for the reference.